File: internal/models/submission.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// ------------------------------------------------------------
// VehicleSubmission – replaces submission:<id> and submissions:* sets
// ------------------------------------------------------------
type VehicleSubmission struct {
	ID string `gorm:"type:varchar(64);primaryKey"`

	EventID   string `gorm:"type:uuid;not null;index:idx_submissions_event"`
	EventSlug string

	ParticipantName  string `gorm:"not null"`
	ParticipantEmail string `gorm:"not null;index:idx_submissions_email"`
	ParticipantPhone string

	VehicleYear          string
	VehicleMake          string
	VehicleModel         string
	VehicleDescription   string
	VehicleModifications string

	Images datatypes.JSON `gorm:"type:jsonb;not null;default:'[]'::jsonb"`

	Status      string    `gorm:"not null;default:pending;index:idx_submissions_status"`
	SubmittedAt time.Time `gorm:"not null"`

	ReviewedAt   *time.Time
	ReviewedBy   string
	ReviewNotes  string

	CheckoutSessionID  string     `gorm:"index:idx_submissions_checkout"`
	CheckoutCreatedAt  *time.Time
	CheckoutCompleted  bool
	CheckoutCompletedAt *time.Time

	PaymentIntentID             string
	PaymentSucceededBeforeApproval bool
	PaymentSucceededAt          *time.Time
	PaymentCaptured             bool
	PaymentCapturedAt           *time.Time

	PriceID       string `gorm:"index:idx_submissions_price"`
	PromotionCode string

	RequiresApproval bool
	AwaitingApproval bool

	ApprovalEmailSent   bool
	ApprovalEmailSentAt *time.Time
	ApprovalEmailResent bool

	TicketID          string     `gorm:"index:idx_submissions_ticket"`
	TicketCreatedAt   *time.Time
	TicketEmailSent   bool
	TicketEmailSentAt *time.Time

	EmailUpdatedAt   *time.Time
	PreviousEmail    string
	EmailResentCount int

	RecoveryAttempts  int        `gorm:"not null;default:0"`
	RecoveryLastSentAt *time.Time

	RefundID        string
	RefundAmount    float64
	RefundIssuedAt  *time.Time

	RevokedAt        *time.Time
	RevokedBy        string
	RevocationReason string

	CreatedAt time.Time `gorm:"autoCreateTime"`
	UpdatedAt time.Time `gorm:"autoUpdateTime"`
}

func (VehicleSubmission) TableName() string {
	return "vehicle_submissions"
}

func (v *VehicleSubmission) BeforeCreate(tx *gorm.DB) error {
	if v.ID == "" {
		id, err := uuid.NewV7()
		if err != nil {
			return err
		}

		v.ID = id.String()
	}

	if v.SubmittedAt.IsZero() {
		v.SubmittedAt = time.Now().UTC()
	}

	return nil
}

type PriceRequirement struct {
	ID string `gorm:"type:uuid;primaryKey" json:"id"`

	PriceID string `gorm:"type:varchar(255);not null;index" json:"price_id"`

	Key      string `gorm:"type:varchar(100);not null" json:"key"`
	Label    string `gorm:"not null" json:"label"`
	Type     string `gorm:"type:varchar(30);not null" json:"type"`
	Required bool   `gorm:"not null;default:false" json:"required"`

	Options datatypes.JSON `gorm:"type:jsonb;not null;default:'[]'::jsonb" json:"options"`

	SortOrder int  `gorm:"not null;default:0" json:"sort_order"`
	Active    bool `gorm:"not null;default:true" json:"active"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`
}

func (PriceRequirement) TableName() string {
	return "price_requirements"
}

func (r *PriceRequirement) BeforeCreate(tx *gorm.DB) error {
	if r.ID == "" {
		id, err := uuid.NewV7()
		if err != nil {
			return err
		}

		r.ID = id.String()
	}

	return nil
}

type SubmissionRequirementAnswer struct {
	ID string `gorm:"type:uuid;primaryKey" json:"id"`

	SubmissionID  string `gorm:"type:varchar(64);not null;index:idx_answers_submission" json:"submission_id"`
	RequirementID string `gorm:"type:uuid;not null;index:idx_answers_requirement" json:"requirement_id"`

	Value datatypes.JSON `gorm:"type:jsonb;not null;index:idx_answers_value" json:"value"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`

	Requirement *PriceRequirement `gorm:"foreignKey:RequirementID;references:ID" json:"requirement,omitempty"`
}

func (SubmissionRequirementAnswer) TableName() string {
	return "submission_requirement_answers"
}

func (a *SubmissionRequirementAnswer) BeforeCreate(tx *gorm.DB) error {
	if a.ID == "" {
		id, err := uuid.NewV7()
		if err != nil {
			return err
		}

		a.ID = id.String()
	}

	return nil
}

// Data Transfer Objects

type VehicleSubmissionDTO struct {
	ID                   string   `json:"id"`
	EventID              string   `json:"event_id"`
	EventSlug            string   `json:"event_slug"`
	ParticipantName      string   `json:"participant_name"`
	ParticipantEmail     string   `json:"participant_email"`
	ParticipantPhone     string   `json:"participant_phone,omitempty"`
	VehicleYear          string   `json:"vehicle_year"`
	VehicleMake          string   `json:"vehicle_make"`
	VehicleModel         string   `json:"vehicle_model"`
	VehicleDescription   string   `json:"vehicle_description,omitempty"`
	VehicleModifications string   `json:"vehicle_modifications,omitempty"`

	Images []string `json:"images"`

	Status      string `json:"status"`
	SubmittedAt string `json:"submitted_at"`

	ReviewedAt  string `json:"reviewed_at,omitempty"`
	ReviewedBy  string `json:"reviewed_by,omitempty"`
	ReviewNotes string `json:"review_notes,omitempty"`

	CheckoutSessionID   string `json:"checkout_session_id,omitempty"`
	CheckoutCreatedAt   string `json:"checkout_created_at,omitempty"`
	CheckoutCompleted   bool   `json:"checkout_completed"`
	CheckoutCompletedAt string `json:"checkout_completed_at,omitempty"`

	PaymentIntentID                string `json:"payment_intent_id,omitempty"`
	PaymentSucceededBeforeApproval bool   `json:"payment_succeeded_before_approval"`
	PaymentSucceededAt             string `json:"payment_succeeded_at,omitempty"`
	PaymentCaptured                bool   `json:"payment_captured"`
	PaymentCapturedAt              string `json:"payment_captured_at,omitempty"`

	PriceID       string `json:"price_id,omitempty"`
	PriceNickname string `json:"price_nickname,omitempty"`
	PromotionCode string `json:"promotion_code,omitempty"`

	RequiresApproval bool `json:"requires_approval"`
	AwaitingApproval bool `json:"awaiting_approval"`

	ApprovalEmailSent   bool   `json:"approval_email_sent"`
	ApprovalEmailSentAt string `json:"approval_email_sent_at,omitempty"`
	ApprovalEmailResent bool   `json:"approval_email_resent"`

	TicketID          string `json:"ticket_id,omitempty"`
	TicketType        string `json:"ticket_type,omitempty"`
	TicketCreatedAt   string `json:"ticket_created_at,omitempty"`
	TicketEmailSent   bool   `json:"ticket_email_sent"`
	TicketEmailSentAt string `json:"ticket_email_sent_at,omitempty"`

	CreatedAt string `json:"created_at"`

	EmailUpdatedAt   string `json:"email_updated_at,omitempty"`
	PreviousEmail    string `json:"previous_email,omitempty"`
	EmailResentCount int    `json:"email_resent_count"`

	RecoveryAttempts  int    `json:"recovery_attempts"`
	RecoveryLastSentAt string `json:"recovery_last_sent_at,omitempty"`

	RefundID       string  `json:"refund_id,omitempty"`
	RefundAmount   float64 `json:"refund_amount,omitempty"`
	RefundIssuedAt string  `json:"refund_issued_at,omitempty"`

	RevokedAt        string `json:"revoked_at,omitempty"`
	RevokedBy        string `json:"revoked_by,omitempty"`
	RevocationReason string `json:"revocation_reason,omitempty"`

	Issues  []string `json:"issues,omitempty"`
	HasIssue bool     `json:"has_issue,omitempty"`

	RequirementAnswers []SubmissionRequirementAnswerDTO `json:"requirement_answers,omitempty"`
}

type SubmissionRequirementAnswerDTO struct {
	ID            string `json:"id"`
	RequirementID string `json:"requirement_id"`
	Key           string `json:"key"`
	Label         string `json:"label"`
	Type          string `json:"type"`
	Value         any    `json:"value"`
}

File: ui/src/lib/schemas/submission.ts
import { z } from "zod";

export const submissionRequirementTypeSchema = z.enum([
	"text",
	"textarea",
	"select",
	"radio",
	"boolean",
	"number",
]);

export const requirementAnswerSchema = z.union([
	z.string(),
	z.number(),
	z.boolean(),
]);

export const submissionRequirementSchema = z.object({
	id: z.string(),
	price_id: z.string(),
	key: z.string(),
	label: z.string(),
	type: submissionRequirementTypeSchema,
	required: z.boolean(),
	options: z.array(z.string()).default([]),
	sort_order: z.number(),
	active: z.boolean(),
});

export const submissionRequirementAnswerSchema = z.object({
	id: z.string(),
	requirement_id: z.string(),
	key: z.string(),
	label: z.string(),
	type: submissionRequirementTypeSchema,
	value: requirementAnswerSchema,
});

export const vehicleSubmissionSchema = z.object({
	id: z.string(),

	event_id: z.string(),
	event_slug: z.string().optional(),

	participant_name: z.string(),
	participant_email: z.email(),
	participant_phone: z.string().optional(),

	vehicle_year: z.string(),
	vehicle_make: z.string(),
	vehicle_model: z.string(),
	vehicle_description: z.string().optional(),
	vehicle_modifications: z.string().optional(),
	images: z.array(z.string()),

	status: z.enum(["pending", "approved", "denied"]),
	submitted_at: z.string(),

	reviewed_at: z.string().optional(),
	reviewed_by: z.string().optional(),
	review_notes: z.string().optional(),

	checkout_session_id: z.string().optional(),
	checkout_created_at: z.string().optional(),
	checkout_completed: z.boolean(),
	checkout_completed_at: z.string().optional(),

	payment_intent_id: z.string().optional(),
	payment_succeeded_before_approval: z.boolean(),
	payment_succeeded_at: z.string().optional(),
	payment_captured: z.boolean(),
	payment_captured_at: z.string().optional(),

	price_id: z.string().optional(),
	price_nickname: z.string().optional(),
	promotion_code: z.string().optional(),

	requires_approval: z.boolean(),
	awaiting_approval: z.boolean(),

	approval_email_sent: z.boolean(),
	approval_email_sent_at: z.string().optional(),
	approval_email_resent: z.boolean(),

	ticket_id: z.string().optional(),
	ticket_created_at: z.string().optional(),
	ticket_email_sent: z.boolean(),
	ticket_email_sent_at: z.string().optional(),

	email_updated_at: z.string().optional(),
	previous_email: z.string().optional(),
	email_resent_count: z.number(),

	recovery_attempts: z.number(),
	recovery_last_sent_at: z.string().optional(),

	refund_id: z.string().optional(),
	refund_amount: z.number(),
	refund_issued_at: z.string().optional(),

	revoked_at: z.string().optional(),
	revoked_by: z.string().optional(),
	revocation_reason: z.string().optional(),

	created_at: z.string(),
	updated_at: z.string(),

	requirement_answers: z
		.array(
			z.object({
				id: z.string(),
				requirement_id: z.string(),
				key: z.string(),
				label: z.string(),
				type: submissionRequirementTypeSchema,
				value: z.unknown(),
			}),
		)
		.optional(),
});

export const issueSubmissionSchema = vehicleSubmissionSchema.extend({
	issues: z.array(z.string()).optional(),
	email_sent: z.boolean().optional(),
});

export type VehicleSubmission = z.infer<typeof vehicleSubmissionSchema>;

export const vehicleSubmissionFormSchema = vehicleSubmissionSchema
	.pick({
		event_id: true,
		participant_name: true,
		participant_email: true,
		participant_phone: true,
		vehicle_year: true,
		vehicle_make: true,
		vehicle_model: true,
		vehicle_description: true,
		vehicle_modifications: true,
		price_id: true,
	})
	.extend({
		participant_name: z.string().trim().min(2, "Enter your full name."),
		participant_email: z.email("Enter a valid email address.").trim(),
		participant_phone: z.string().trim().optional(),
		vehicle_year: z
			.string()
			.trim()
			.regex(/^\d{4}$/, "Enter a four-digit year."),
		vehicle_make: z.string().trim().min(2, "Enter the vehicle make."),
		vehicle_model: z.string().trim().min(2, "Enter the vehicle model."),
		vehicle_description: z.string().trim().optional(),
		vehicle_modifications: z.string().optional(),
		event_id: z.string().trim().min(1),
		price_id: z.string().trim().min(1),
	});

export type VehicleSubmissionFormData = z.infer<
	typeof vehicleSubmissionFormSchema
>;

export type SubmissionRequirementType = z.infer<
	typeof submissionRequirementTypeSchema
>;

export type RequirementAnswer = z.infer<typeof requirementAnswerSchema>;

export type SubmissionRequirement = z.infer<typeof submissionRequirementSchema>;

export type SubmissionRequirementAnswer = z.infer<
	typeof submissionRequirementAnswerSchema
>;

export type IssueSubmission = z.infer<typeof issueSubmissionSchema>;
